### ISSUE TITLE
Feature : Add write feature in redshift through spark-core and spark-sql

### ISSUE THREAD

**Issue by mkanchwala** [databricks/spark-redshift#8]
There should be a way to write the RDDs to Redshfit in Spark with some example

**Comment by mengxr** [comment]
We are going to implement the data source API with read/write support. I'll keep you updated.

**Comment by Aerlinger** [comment]
Any updates on this? I'm looking to implement this myself and would rather join forces than duplicate work. Happy to help out in any way I can.

**Comment by bumrush** [comment]
I would like to know about this too. Is there a way to write RDDs to Redshift?

**Comment by mengxr** [comment]
Sorry, we haven't had time to implement this feature. The current plan is to make our design doc public and see if there are people who are interested in working on it. I will add the link soon.

**Comment by jaley** [comment]
Hi there!

We (Analytics team at SwiftKey) have planned some capacity in our next sprint to get our Spark jobs writing data out to Redshift. Were were planning to do something crude with an S3 CSV export, but the better way for it to work would probably be by adding the write support to this library. If you're able to provide some support, perhaps on the mailing list, this thread, IRC, or whatever you prefer to use, we'd be happy to have a go at implementing this feature and contributing the code back.

Are you able to share that design document you mentioned? We're likely to need a bit of guidance, and that would be a good start.

Also, we have a bunch of things that we'd like this feature to involve to make our ETL pipeline work the way we need. Hopefully these are generally desirable features for the write functionality, but here's our wish list, which we'd hope to include in the implementation, for wider review:

* DataFrames can be saved to a Redshift table via `save()` function (think that's pretty core!)
* Automatically infers a 'default' Redshift schema from the DataFrame column types, but it should be possible to provide more specific overrides for these with options somehow. For example, a DataFrame `StringType` column may sometimes needed to be written as `TIMESTAMP WITHOUT TIMEZONE NOT NULL` in Redshift.
* Ability to specify distribution key, sort keys, etc.
* Data load should be done via S3, with potentially configurable, but at least sensible default data encodings, compression, etc. We could potentially make use of the new Avro support in Redshift, if there are performance advantages (I'm not sure). We don't actually require support for loading data via SQL `INSERT` commands, and I imagine most people don't, as it's painfully slow and Amazon don't recommend it.
* We primarily want to drop entire tables and replace them in the ETL pipeline. This is because we favour batch recomputes of entire "views" of our data wherever possible, and it works nicely with Redshift's compression and sorting performance characteristics (as Redshift doesn't like updates), so we'll want to focus on making that use case easy, though incremental updates shouldn't be hard to support either.
* It would also be nice if that big truncate and swap of the data happened via a staging table first. Our experience has been that a common point of failure in Redshift-related ETL job is that the final load is a common point of failure, as there's no easy way to predict whether the data is strictly valid according to the Redshift schema. Leaving out-of-date data in the table is generally preferred to dropping and then failing for us!

How does this align with the design that you guys already had in mind? If you think it makes sense for us to work on this, we'd like to get started fairly soon (in the next week or so) - is that realistic?

Many thanks!

**Comment by mengxr** [comment]
@jaley @Aerlinger @bumrush  This is the design doc and it would be great if you have time to work on it:

https://docs.google.com/a/databricks.com/document/d/1JT3jczx_jTXbxCCKG5EAoMF3WPtlosccs2J4s7kxl0E/edit?usp=sharing

@jaley I'm not aware of the Avro support in Redshift. Could you provide the link? Avro is certainly better than handling multi-line records:)

As the first step, I think we should implement the data source API for Redshift using JDBC and S3 with basic functions (read/write). Then we can consider further optimizations. Feel free to comment on the design doc or reply on this thread. It would be fantastic to have this implemented.

**Comment by Aerlinger** [comment]
Thanks @mengxr, looking forward to reading the design doc. I've requested access with an email consistent with my username.

@jaley The team I'm working with has been developing an ETL pipeline from Postgres -> Redshift and we've faced similar problems/approaches to what you're describing, although the trickiest issue has been dealing with datatype inference/conversion and managing schema changes. With regards to making table updates (nearly) atomic, have you considered loading the fresh data into a staging table, then renaming the original table to a temporary table name, followed by a final rename of the staging table to the original table name?

i.e. something like this:

```SQL
ALTER TABLE RENAME original_table TO original_archive_<timestamp_of_last record>
ALTER TABLE RENAME staging_table TO original_table
```

I came across a blog article with a similar approach:

https://www.simple.com/engineering/safe-migrations-with-redshift

**Comment by mengxr** [comment]
@Aerlinger Sorry I forgot to set the permission. Please retry that link.

**Comment by jaley** [comment]
@mengxr - thanks for sharing the doc!

@Aerlinger - Yes, that rename strategy was exactly what I had in mind for staging some data temporarily, making sure the load succeeds before swapping. I guess it'll have to be configurable, as it requires extra disk space in Redshift.

Avro support in Redshift was mentioned at AWS Summit in London at the end of May, and we requested preview access last week, so I'll update this thread with news on the specifics when I have access. If it's acceptable for this library to have a dependency on spark-avro, I think it might make for a sensible 'default' serialization format for the S3-based data exchange proposed in the design doc. In addition to the likely performance benefits, it'll avoid escape character issues with the text based formats, which I've always found to be a pain-point when round-tripping data to and from Redshift using `UNLOAD` and `COPY`.

One minor thing to note -- I think Redshift can also `COPY` from HDFS on an EMR cluster, which may be useful to some. I guess the abstractions already in Spark for S3 and HDFS pretty much already take care of that, so it might 'just work'? Unfortunately I don't think it's supported for `UNLOAD` commands.

I'm hoping to get started on this today and tomorrow, so I'll no doubt be in touch with questions etc.

**Comment by jaley** [comment]
I made a start at this today and made some progress. I have a basic `RedshiftRelation` implementation talking to Redshift over JDBC to get the schema etc. I have a couple of things I wouldn't mind a second opinion on if someone has a moment though!

* What do we want the read API looking like? I was initially tempted to have the caller specify a table name, for example, `sqlContext.redshiftTable('my_table', jdbcUrl, ...)`, but then I realised that this is kind of limiting. Could it be that sometimes it makes sense to run a query involving some joins on the Redshift side, before loading the results into a DataFrame for further processing in Spark? Conversely, if you don't require a table name, would would the behaviour of `buildScan()` be? Looking at most of the optional traits, it really seems to make sense that `RedshiftRelation` applies to one single table. Perhaps it would make sense to provide a `RedshiftTableRelation`, for use cases where the caller wants to do all of the work in Spark, and a `RedshiftQueryRelation`, where the caller wants to seed the DataFrame with the output of Redshift query? In both cases, it should be possible to infer the schema using the JDBC APIs. Any suggestions here?

* The JDBC driver dependency situation is a bit ugly. Amazon no-longer recommend using the Postgres driver, and instead provide their own JDBC4 driver, but this isn't provided as a Maven/Ivy dependency anywhere useful, as far as I can see. Amazon provide a binary [here](http://docs.aws.amazon.com/redshift/latest/mgmt/configure-jdbc-connection.html), but note that there's also some licensing/redistribution requirements. I've been testing with a direct dependency on an old Postgres 8.3 driver so far, but perhaps we should assume a JDBC driver will be provided and allow users to add the official jar themselves?

* Any thoughts on serialisation? I'm talking to AWS to see when we can expect the Avro support they announced in May, as I'm hoping it'll simplify the whole process by allowing us to make use of the schema Redshift embeds in the data itself. Failing that, though it would be nice to make use of the existing `RedshiftInputFormat` class in the library, I'm not sure it makes sense for us to implement a serializer to write data back out to S3 in that format when adding write support? Perhaps we could just use some commodity format like TSV/CSV?

Thanks in advance for any suggestions / guidance. I should be back on this from Wednesday - out of the office tomorrow.

**Comment by jaley** [comment]
FYI - I just noticed that the Avro support for Redshift `COPY` is now in the reference docs:
http://docs.aws.amazon.com/redshift/latest/dg/r_COPY.html

Seems it's not in `UNLOAD`, so we can't use it for a full roundtrip :-(

**Comment by jaley** [comment]
Hey guys,

I've made some decent progress on this. I have a WIP branch on my fork, which you can try out here if you're interested:
https://github.com/jaley/spark-redshift/tree/jdbc-avro

What I've done so far:

* Started plugging in Data Source API implementation - much more can be done though.
* Simple TableScan implementation to load all data from a table using JDBC to trigger an unload to S3, then make use of the existing file format code to read it into a DataFrame.
* I realised some type conversions are needed, because everything is parsed as a String from the unloaded file. We can use the schema as parsed by the JDBC source to perform necessary conversions though, I've done this for Timestamps as a test so far.
* Fairly naive write implementation, using Avro as the intermediate file format in S3, again triggering COPY via JDBC. The writing out to S3 works, but I need my dev Redshift cluster to by patched to the latest version for the Avro support - that'll happen over the weekend, so I'll resume on Monday!

There are a few things I could use some help / suggestions on if anyone has time to offer some advice:

* Dependency hell. I nearly threw my laptop across the room. See the `build.sbt` file where I have a bunch of TODOs. I can't make things play nicely with Avro, Hadoop 2, Spark 1.4, etc. Keep getting problems with javax-servlet and jets3t conflicts.
* The JDBC functionality in `spark-sql` is package-private. Anybody know why this is? There's some really useful code in there that I'd hate to reimplement (schema conversion stuff mostly). I have a nasty shim in the Spark SQL package to access it right now, but I'd welcome better suggestions.
* I'm planning to add the other Data Source API traits to `RedshiftRelation`, as filtering and inserting should be easily doable.
* The JDBC driver dependency isn't the one Amazon recommends. Maybe we should make this a `provided` dependency? I'm not sure what to do, as the library is broken without a driver, but there are probably licensing complications around redistributing Amazon's driver?
* It would be good to expose more configuration for the Redshift functionality. Particularly overwrite behaviours, using staging tables, adding sort and distribution keys to the schema etc. Feels like this is going to be a bit awkward. I'll try to respect `SaveMode` and do the "sensible" thing, I guess.
* Testing, of which there is currently none.

Let me know if anyone has suggestions / ideas. Also, I'm happy to break this up and get the basic functionality in first. I think I'm pretty close to having a basic data source implementation with full read/write, albeit a bit inflexible, and maybe that'd be a good starting point?

**Comment by jaley** [comment]
@mengxr Official Avro announcement just went up: https://aws.amazon.com/about-aws/whats-new/2015/07/amazon-redshift-now-supports-avro-ingestion/

My branch now has a full data source implementation, btw. I'm hoping to get it cleaned up and send over a PR later this week.

**Comment by mengxr** [comment]
@jaley Thanks for sending the PR and sorry for my slow response! Let's continue our discussion to the PR page. I will ping @kai-zeng to review your PR, who is more familiar with the data source API.

Since Redshift only support COPY from Avro files, we still need to UNLOAD in text format and read them into Spark using the input format implemented. Adding Avro support in the data source API might not be able to reduce the code complexity. Did you test performance? Well, let's discuss more on the PR page.

**Comment by marmbrus** [comment]
Marking this closed now that #22 is merged.
